c++/src/kj/common.h

```diff
@@ -294,7 +294,10 @@ KJ_NORETURN(void unreachable());
 }  // namespace _ (private)
 
 #ifdef KJ_DEBUG
-#if _MSC_VER && !defined(__clang__)
+#ifndef _MSVC_TRADITIONAL
```

Comment:
Could use a comment about that this is comparability support for Visual Studio < 2017 version 15.8.

I think a more elegant way that doesn’t run afoul of reserved identifiers is to do something like:
```
#define KJ_MSVC_TRADITIONAL_CPP (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)
```

Then you can just use `KJ_MSVC_TRADITIONAL_CPP`.

Reply:
I tried this:

```
#define KJ_MSVC_TRADITIONAL_CPP (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)
#if _MSC_VER && !defined(__clang__) && KJ_MSVC_TRADITIONAL_CPP
```

It turns out clean in MSVC and Clang compilations, but the Qt Creator IDE reports the following warning:

```
-Wexpansion-to-defined:
common.h:298:40: warning: macro expansion producing 'defined' has undefined behavior
common.h:297:35: note: expanded from macro 'KJ_MSVC_TRADITIONAL_CPP'
```

Reply:
Sounds like you probably need to do:

```
#if _MSC_VER && (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)
#define KJ_MSVC_TRADITIONAL_CPP 1
#endif
```

Reply:
Why does your `#if` condition include `_MSC_VER` but not `!defined(__clang__)`? IMO I'd either remove the `_MSC_VER` check (this approach sets `KJ_MSVC_TRADITIONAL_CPP=1` even on non-MSVC compilers, but that should be fine if nobody checks it):

```cpp
#if !defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL
#define KJ_MSVC_TRADITIONAL_CPP 1
#endif
#if _MSC_VER && !defined(__clang__) && KJ_MSVC_TRADITIONAL_CPP
```

Or add the `!defined(__clang__)` check:

```cpp
#if _MSC_VER && !defined(__clang__) && (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)
#define KJ_MSVC_TRADITIONAL_CPP 1
#endif
#if KJ_MSVC_TRADITIONAL_CPP
```

At which point the one-liner is simpler, and better since it doesn't leak a #define (though worse since it doesn't give the expression a name):

```cpp
#if _MSC_VER && !defined(__clang__) && (!defined(_MSVC_TRADITIONAL) || _MSVC_TRADITIONAL)
```

Reply:
I forgot about clang. Yes that should probably be in there too.

It seemed easier to define `KJ_MSVC_TRADITIONAL_CPP` once and then use that everywhere rather than have a multi-part check at every use site.

Reply:
In the part of capnproto I build in my vendored project, this header is the only place I had to fix to make it build on conformant MSVC. Honestly I feel it's not worth the trouble to make capnproto's source files build properly on conformant MSVC.

## My attempt to make it build

When building the entire capnproto project on MSVC with `add_compile_options(/Zc:preprocessor)`, I found that debug.h and test.h also depended on the preprocessor. Both files already see `KJ_MSVC_TRADITIONAL_CPP` when needed, but neither directly includes common.h; should I add an explicit `#include "common.h"`?

Worse yet, the GCC branch of `KJ_ASSERT_NONNULL`/`KJ_REQUIRE_NONNULL` fails to build on MSVC due to using `({...})` statement expressions, and the MSVC branch fails to build on MSVC due to the preprocessor differences. How should this be handled?

Assuming you don't want 3 code paths, one possibility is to stop using statement expressions in the GCC path, and switch conformant MSVC to the GCC path. I implemented this change locally, and got the entire capnproto to compile, except for a single error where since debug.h takes the GCC branch, it triggers the "MSVC mis-deduces `ContextImpl<decltype(func)>` as `ContextImpl<int>`" bug again... 😢

Another is to not support building `KJ_REQUIRE_NONNULL` and `KJ_ASSERT_NONNULL` in conformant MSVC (debug.h is exported to clients through "capnp/message.h", but these macros are never expanded in public headers, only private source files). Then when all supported MSVC versions support the conformant preprocessor, switch the MSVC codepath to the conformant preprocessor but keep MSVC hacks as necessary. However this may have issues if you ever want to support both MSVC versions that lack the new preprocessor, and MSVC versions that lack the old one.

## Compiler tiers?

Is capnproto-c++'s main testing environment Linux GCC (and perhaps Clang), with Windows MSVC and Clang as less-supported, less-tested targets that frequently run into build issues?

Reply:
Every file includes `kj/common.h`, though sometimes it may be transitively through other headers.

I think we should be able to change `({...})` to IIFE lambdas if necessary.

`debug.h` is intended to be useful for any project that uses KJ, so we will need to fix it at some point. But I won't force you to do it if you don't want to. I would prefer, though, that `KJ_MSVC_TRADITIONAL_CPP` is only true when we're actually using MSVC, otherwise I think the macro is confusing. This should be a minor change to your PR.

Reply:
Force pushed. Hopefully it's good enough for now.